Claimdrops Smart Contract

ali-h opened the pull request:
Claimdrops Smart Contract, can be used to create coin offering with custom prices. creates a pool which has a expiration time upto 90 days (variable) and maximum claims (variable). anyone or selected users can claim tokens for the defined price in Hive Pegged token (similar to market). there can be a claim limit for everyone or individuals in a list which can go up than the pool itself so its first to fill the pool. the price collected, and pool (if remaining) after the expiration is sent to a account or contract defined by owner.

Actions:
* create: creates a claimdrop
fee 50 BEE and 0.1*maxClaims is burned.
* claim: can claim tokens from an active claimdrop
price in SWAP.HIVE is sent to the owner, it is calculated by quantity parameter. price*quantity (price defined in the claimdrop)

donchate commented:
A nice option would be to distribute unclaimed pool amount to those who claimed.

bt-cryptomancer commented on contracts/claimdrops.js:
```diff
@@ -0,0 +1,346 @@
+      && api.assert(api.isValidAccountName(account), `list[${i}]: invalid account name`)
+      && api.assert(limit, `list[${i}]: limit cannot be undefined`)
+      && api.assert(!api.BigNumber(limit).isNaN(), `list[${i}]: invalid limit`)
+      && api.assert(api.BigNumber(limit).gt(0), `list[${i}]: limit must be positive`)
```
Can you do a max size check for limit?  Can be something like asserting that api.BigNumber(limit).lte(Number.MAX_SAFE_INTEGER) and also use the BigNumber isFinite function - https://mikemcl.github.io/bignumber.js/#isF

bt-cryptomancer commented on contracts/claimdrops.js:
```diff
@@ -0,0 +1,346 @@
+    && api.assert(symbol && typeof symbol === 'string'
+      && price && typeof price === 'string' && !api.BigNumber(price).isNaN()
+      && pool && typeof pool === 'string' && !api.BigNumber(pool).isNaN()
+      && maxClaims && Number.isInteger(maxClaims)
```
Can you do a max size check for maxClaims?  Should make sure it is less than Number.MAX_SAFE_INTEGER

bt-cryptomancer commented on contracts/claimdrops.js:
```diff
@@ -0,0 +1,346 @@
+            claimdrop.list = parsedList;
+          } else return;
+        } else if (limit) {
+          if (api.assert(api.BigNumber(limit).gt(0), 'limit must be positive')
```
Can you do a max size check for limit? Can be something like asserting that api.BigNumber(limit).lte(Number.MAX_SAFE_INTEGER) and also use the BigNumber isFinite function - https://mikemcl.github.io/bignumber.js/#isF

bt-cryptomancer reviewed (changes requested):
This mostly looks good.  I think it would be good to have a bit stricter validation on max size limits, see comments at indicated code sections.

I'd also like to avoid adding another virtual transaction to Block.js, as we already have several of those and I don't want to add any more unless absolutely necessary.  Let's have checkExpiredClaimdrops be called first whenever someone uses the claim action.  And it would be good to have an expire action that only the claimdrop owner can use, so the owner can manually trigger expiration when desired, instead of waiting for the next claim action to do it.

bt-cryptomancer commented:
> A nice option to add would be to distribute unclaimed pool amount to those who claimed.

Agreed, I don't think this is necessary for the first version of the contract, but would be a cool feature to add in a version 2 release.

ali-h commented:
> A nice option to add would be to distribute unclaimed pool amount to those who claimed.

a nice feature, this can surely be added in the next version as an optional choice for the owner.

ali-h commented:
> This mostly looks good. I think it would be good to have a bit stricter validation on max size limits, see comments at indicated code sections.
> 
> I'd also like to avoid adding another virtual transaction to Block.js, as we already have several of those and I don't want to add any more unless absolutely necessary. Let's have checkExpiredClaimdrops be called first whenever someone uses the claim action. And it would be good to have an expire action that only the claimdrop owner can use, so the owner can manually trigger expiration when desired, instead of waiting for the next claim action to do it.

I've added stricter validations so its less than MAX_SAFE_INTEGER but is `isFinite` necessary? if the value is less than safe integer its finite right.

Yes, initially I was going to add expireClaimdrops in the claim action, but thought it might get the processing slow? but its probably better than adding virtual transaction every block for it. so not searching for all expired claimdrops, now when a claim action is called, if the requested claimdrop has reached its expiration time it gets deleted (the expire action triggers).

I've modified `checkExpiredClaimdrops`  to be `expire` which can be called by the sender or owner set by the sender of a claimdrop, this gives the option to expire a claimdrop manually anytime without depending on claim action.

I've added sender and owner both to the claimdrop object because maybe if the owner is a contract it can not call expire event.

bt-cryptomancer reviewed (changes requested):
> I've added stricter validations so its less than MAX_SAFE_INTEGER but is isFinite necessary? if the value is less than safe integer its finite right.

eonwarped started using isFinite in his contract code, which I thought was a nice idea, but you're right that's a redundant check since you're comparing with Number.MAX_SAFE_INTEGER anyway.  So this is fine.

> Yes, initially I was going to add expireClaimdrops in the claim action, but thought it might get the processing slow? but its probably better than adding virtual transaction every block for it. so not searching for all expired claimdrops, now when a claim action is called, if the requested claimdrop has reached its expiration time it gets deleted (the expire action triggers).

Expiring a claimdrop is a pretty cheap operation, so I'd still like to see a claim operation expire all old claimdrops, not just the one being claimed.  The market contract works similarly; when a market order is placed, it removes all old orders which have already expired.

All other changes look fine.

bt-cryptomancer reviewed (changes requested):
The way you've got the removeExpiredClaimdrops function coded now introduces the possibility of non-deterministic behavior.  Although this is extremely unlikely with this kind of contract, it's still something that needs to be handled properly.  The problem is that api.db.find implicitly limits query results to 1000 when you don't explicitly provide the limit parameter.  Thus if there are more than 1000 expired claimdrops, this code won't work as intended (and without using indexes on the query, the order of query results is not well defined).

I see 2 options to fix this problem:

1.  Add back in the maxExpiresPerBlock setting that you originally had, as well as the use of the _id index.

OR

2. Wrap the query in a while loop to ensure the case of > 1000 query results is handled OK.  For an example of this technique, see the actions.checkPendingUndelegations function in nft.js or upgradeUserSchema function in botcontroller.js

bt-cryptomancer commented:
Of the two options I suggested to resolve the query issue, if using number 2 I think it's still important to make sure an index is used for sorting.  We probably should do an audit of the other contracts at some point and ensure indexes are always being used, or else change the core node software to always add the _id index if one is not specified.

ali-h commented:
changes made. I did think about this but got up with the likely scenario. but yup contracts can be abused if there are no checks for things like this. I've added similar loop to `checkPendingUndelegations` in nft contract where loop continues if there are more than 1000 objects in the db with sorting by _id.

bt-cryptomancer reviewed (approved):
changes look good
